Universal/Sniffs/CodeAnalysis/MixedBooleanOperatorSniff.php

```diff
@@ -0,0 +1,107 @@
+ *
+ * @link https://github.com/squizlabs/PHP_CodeSniffer/pull/3205
+ * @link https://github.com/php-fig/per-coding-style/issues/22
+ *
+ * @since 1.2.0
+ */
+final class MixedBooleanOperatorSniff implements Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 1.2.0
+     *
+     * @return array<int|string>
+     */
+    public function register()
+    {
+        return [
+            \T_BOOLEAN_OR,
+            \T_BOOLEAN_AND,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 1.2.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token
+     *                                               in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        $token  = $tokens[$stackPtr];
+
+        $start = $phpcsFile->findStartOfStatement($stackPtr);
+
+        if ($token['code'] === T_BOOLEAN_AND) {
+            $search = T_BOOLEAN_OR;
+        } elseif ($token['code'] === T_BOOLEAN_OR) {
+            $search = T_BOOLEAN_AND;
+        } else {
+            throw new \LogicException('Unreachable');
+        }
+
+        while (true) {
+            $previous = $phpcsFile->findPrevious(
+                [
+                    $search,
+                    T_OPEN_PARENTHESIS,
+                    T_OPEN_SQUARE_BRACKET,
+                    T_CLOSE_PARENTHESIS,
+                    T_CLOSE_SQUARE_BRACKET,
+                ],
+                $stackPtr,
+                $start
+            );
+
+            if ($previous === false) {
+                break;
+            }
+
+            if ($tokens[$previous]['code'] === T_OPEN_PARENTHESIS
+                || $tokens[$previous]['code'] === T_OPEN_SQUARE_BRACKET
+            ) {
+                // We halt if we reach the opening parens / bracket of the boolean operator.
+                return;
+            } elseif ($tokens[$previous]['code'] === T_CLOSE_PARENTHESIS) {
+                // We skip the content of nested parens.
+                $stackPtr = ($tokens[$previous]['parenthesis_opener'] - 1);
+            } elseif ($tokens[$previous]['code'] === T_CLOSE_SQUARE_BRACKET) {
+                // We skip the content of nested brackets.
+                $stackPtr = ($tokens[$previous]['bracket_opener'] - 1);
+            } elseif ($tokens[$previous]['code'] === $search) {
+                // We reached a mismatching operator, thus we must report the error.
+                $error = "Mixed '&&' and '||' within an expression without using parentheses.";
+                $phpcsFile->addError($error, $stackPtr, 'MissingParentheses', []);
+                return;
+            } else {
+                throw new \LogicException('Unreachable');
```

Comment:
As indicated by the exception message, I assume the `else` branch to never be reached. If it would be reached, it would indicate a programming error on my part.

In my code I've adopted this pattern to make sure any such programming error becomes immediately apparent, instead of the code silently misbehaving.

That said: I can completely remove the `else` branch if that is what you prefer.

Reply:
I understand your reasoning, but if you haven't been able to come up with (test) code which would trigger this, this will never be triggered in a dev situation, so the only time the Exception gets thrown is at runtime for end-users, which causes a problem in _their_ code scans/CI as the file under scan doesn't get scanned completely anymore.

You could leave the `else`, but replace the `throw ...` statement with a `return`, which will not cause a problem for end-users.

In those cases, I normally accompany the `return` with a `// Shouldn't be possible, but just in case.` comment + a `// @codeCoverageIgnore`.

Reply:
Okay, removed.